components/feature/prompts/src/main/java/mozilla/components/feature/prompts/logins/LoginValidationDelegate.kt

```diff
@@ -0,0 +1,101 @@
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.prompts.logins
+
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers.IO
+import kotlinx.coroutines.async
+import mozilla.appservices.logins.InvalidLoginReason
+import mozilla.appservices.logins.InvalidRecordException
+import mozilla.appservices.logins.ServerPassword
+import mozilla.components.concept.engine.autofill.Login
+import mozilla.components.feature.prompts.logins.LoginValidationDelegate.Result
+import mozilla.components.service.sync.logins.AsyncLoginsStorage
+
+/**
+ * Provides a method for checking whether or not a given login can be stored.
+ */
+interface LoginValidationDelegate {
+
+    sealed class Result {
+        /**
+         * Indicates that the [Login] does not currently exist in the storage, and a new entry
+         * with its information can be made.
+         */
+        object CanBeCreated : Result()
+        /**
+         * Indicates that a matching [Login] was found in storage, and the [Login] can be used
+         * to update its information.
+         */
+        object CanBeUpdated : Result()
+        sealed class Error : Result() {
+            /**
+             * The passed [Login] had an empty password field, and so cannot be saved.
+             */
+            object EmptyPassword : Error()
+            object NotImplemented : Error()
+            data class GeckoError(val exception: InvalidLoginReason) : Error()
+        }
+    }
+
+    fun validateCanPersist(login: Login): Deferred<Result>
+}
+
+/**
+ * Default [LoginValidationDelegate] implementation that always returns false.
+ *
+ * This can be used by any consumer that does not want to make use of autofill APIs.
+ */
+class NoopLoginValidationDelegate : LoginValidationDelegate {
+    override fun validateCanPersist(login: Login): Deferred<Result> {
+        return CompletableDeferred(Result.Error.NotImplemented)
+    }
+}
+
+/**
+ * A delegate that will check against [storage] to see if a given Login can be persisted, and return
+ * information about why it can or cannot.
+ */
+class DefaultLoginValidationDelegate(
+    private val storage: AsyncLoginsStorage,
+    private val passwordsKey: String
+) : LoginValidationDelegate {
+    override fun validateCanPersist(login: Login): Deferred<Result> {
+        try {
+            return CoroutineScope(IO).async {
+                storage.ensureUnlocked(passwordsKey).await()
+                // TODO this should share logic from mozilla.components.browser.engine.gecko.autofill.Login.toServerPassword,
+                //  but it can't depend on GV. 'service-sync-logins' and 'concept-engine' have no good shared descendant
+                //  where this can live, where can it be moved?
+                storage.ensureValid(ServerPassword(
+                    id = login.guid ?: "",
```

Comment:
I'd also love some help on this TODO too.  The conversions between GV/AS/AC representations of logins are simple, but I'd like to avoid littering the code with them.  Any ideas where this could live?

Reply:
The TODO is cut off in the comment.  Full text:

`TODO this should share logic from mozilla.components.browser.engine.gecko.autofill.Login.toServerPassword, but it can't depend on GV. 'service-sync-logins' and 'concept-engine' have no good shared descendant where this can live, where can it be moved?`